paddle/fluid/memory/allocation/cuda_malloc_async_allocator.h

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <mutex>  // NOLINT
+#include <unordered_set>
+
+#include "paddle/fluid/memory/allocation/allocator.h"
+#include "paddle/fluid/memory/allocation/spin_lock.h"
+#include "paddle/fluid/platform/place.h"
+
+namespace paddle {
+namespace memory {
+namespace allocation {
+
+// TODO(eee4017): It may be beneficial to introduce an abstract class named
+// `StreamAllocator` in future developments. This class would serve as a central
+// entity for methods specifically related to stream management, such as
+// `RecordStream` and `EraseStream`. The introduction of `StreamAllocator` would
+// enable both `StreamSafeCUDAAllocator` and `CUDAMallocAsyncAllocator` to
+// inherit directly from it,
+
+// The `CUDAMallocAsyncAllocation` class extends `Allocation` and is used for
+// managing memory allocations with CUDA async malloc. It includes methods to
+// handle stream associations and to query the owning stream of the allocation.
+class CUDAMallocAsyncAllocation : public Allocation {
+ public:
+  CUDAMallocAsyncAllocation(void* ptr,
+                            size_t size,
+                            platform::Place place,
+                            gpuStream_t stream)
+      : Allocation(ptr, size, place),
+        malloc_stream_(stream),
+        used_in_another_stream(false) {}
+
+  gpuStream_t GetOwningStream() const { return malloc_stream_; }
+
+  // TODO(eee4017): The current implementation of RecordStream is
+  // similar to that in StreamSafeCUDAAllocator. This approach might lead to
+  // host execution blocking and redundant EventQuery checks. Considering
+  // cudaMallocFree, stream-ordered semantics could be leveraged for more
+  // efficient device-side release.
+  void RecordStream(gpuStream_t stream);
+  void RecordGraphCapturingStreams();
+  void RecordStreamWithNoGraphCapturing(gpuStream_t stream);
+  void EraseStream(gpuStream_t stream);
+  bool CanBeFreed(bool synchronize = false);
+  void Free(int dev_id);
+
+ private:
+  static thread_local std::once_flag once_flag_;
+  gpuStream_t malloc_stream_;
+  bool used_in_another_stream;
+  std::set<gpuStream_t> graph_capturing_stream_set_;
+  SpinLock event_map_lock_;
+  std::map<gpuStream_t, gpuEvent_t> event_map_;
+};
+
+// The `CUDAMallocAsyncAllocator` class extends `Allocator` and is specialized
```

Comment:
When introducing CUDAMallocAsyncAllocator with stream-ordered semantics, why do we still need a complex CUDA event releated mechanism similar to StreamSafeCUDAAllocator involving EventRecord, EventQuery and unfreed Allocation management? Since in Paddle the Allocator malloc and free are dispatched to the same stream as their relevant OP kernel, and the cross-stream synchronization for kernel is guaranteed by upstream code, can we completely transfer all management responsibilities to CUDA, and just do some simple CUDAMalloc/FreeAsync calls in Allocator?

Reply:
The development of a mechanism designed to fully offload memory management responsibilities to CUDA is currently in progress, utilizing memory streams for this specific aim. This initiative remains under construction and could see significant advancements in forthcoming pull requests. At present, the necessity for employing APIs akin to EventRecord, EventQuery, and the management of unfreed allocations persists, primarily because the stream passed into the Allocator defaults to the CUDA stream, and the Record stream is employed to mark the specific stream that the memory block operates on. Our ongoing efforts are focused on addressing these issues, with a commitment to refining and enhancing this aspect in future updates.